File: Olympiad/JOI/official/JOIOC/2018/code/2018-open-collapse.pas
unit collapse;

interface

type
  ArrayOfLongint = array of longint;

function simulateCollapse(N: longint; T_, X_, Y_, W_, P_: ArrayOfLongint): ArrayOfLongint;

implementation

uses math;

type Compare=function(var a,b:longint):boolean;
function lowerBound(var v:array of longint;l,r:longint;a:longint;c:Compare):longint; var i:longint; begin l-=1; while l+1<r do begin i:=(l+r)div 2; if c(v[i],a) then l:=i else r:=i; end; exit(r); end;
procedure swap(var a,b:longint); var t:longint; begin t:=a; a:=b; b:=t; end;
procedure sort(var v:array of longint;l,r:longint;c:Compare); var i,j,k:longint; begin while r-l>4 do begin i:=l+1; j:=(l+r)div 2; k:=r-1; if c(v[i],v[j]) then if c(v[j],v[k]) then swap(v[l],v[j]) else if c(v[i],v[k]) then swap(v[l],v[k]) else swap(v[l],v[i]) else if c(v[i],v[k]) then swap(v[l],v[i]) else if c(v[j],v[k]) then swap(v[l],v[k]) else swap(v[l],v[j]); i:=l+1; j:=r-1; while true do begin while c(v[i],v[l]) do i+=1; while c(v[l],v[j]) do j-=1; if i>=j then break; swap(v[i],v[j]); i+=1; j-=1; end; sort(v,i,r,c); r:=i; end; for i:=0 to r-l-2 do for j:=l to r-2 do if c(v[j+1],v[j]) then swap(v[j],v[j+1]); end;

const
  BLOCK_SIZE: longint = 333;

var
	C, Q: longint;
  T, X, Y, W, P: array of longint;
  uf: array of longint;
  historyLen: longint;
  historyKey, historyVal: array [0 .. 1000010] of longint;

function cmpEdge(var i, j: longint): boolean;
begin
  if Y[i] <> Y[j] then exit(Y[i] < Y[j]) else if X[i] <> X[j] then exit(X[i] < X[j]) else exit(i < j);
end;
function cmpQuery(var i, j: longint): boolean;
begin
  if P[i] <> P[j] then exit(P[i] < P[j]) else exit(i < j);
end;

function root(u: longint): longint;
begin
  if uf[u] < 0 then begin
    exit(u);
  end else begin
    uf[u] := root(uf[u]);
    exit(uf[u]);
  end;
end;
function connect(u, v: longint): boolean;
begin
  u := root(u);
  v := root(v);
  if u = v then exit(false);
  if uf[u] > uf[v] then swap(u, v);
  uf[u] += uf[v];
  uf[v] := u;
  exit(true);
end;

procedure assignSaving(key, val: longint);
begin
  if uf[key] <> val then begin
    historyKey[historyLen] := key;
    historyVal[historyLen] := uf[key];
    historyLen += 1;
    uf[key] := val;
  end;
end;
function rootSaving(u: longint): longint;
begin
  if uf[u] < 0 then begin
    exit(u);
  end else begin
    assignSaving(u, rootSaving(uf[u]));
    exit(uf[u]);
  end;
end;
function connectSaving(u, v: longint): boolean;
begin
  u := rootSaving(u);
  v := rootSaving(v);
  if u = v then exit(false);
  if uf[u] > uf[v] then swap(u, v);
  assignSaving(u, uf[u] + uf[v]);
  assignSaving(v, u);
  exit(true);
end;

function simulateCollapse(N: longint; T_, X_, Y_, W_, P_: ArrayOfLongint): ArrayOfLongint;
var
  ans: array of longint;
  edges, queries: array of longint;
  edgesLen, queriesLen: longint;
  link: array of longint;
  cntGlobal, cntLocal: longint;
  e, f, g, h, i, j, k, u, phase: longint;
begin
	C := Length(T_);
	Q := Length(W_);
  setlength(T, C);
  setlength(X, C);
  setlength(Y, C);
  for i := 0 to C - 1 do begin
    T[i] := T_[i];
    X[i] := X_[i];
    Y[i] := Y_[i];
  end;
  setlength(W, Q);
  setlength(P, Q);
  for h := 0 to Q - 1 do begin
    W[h] := W_[h];
    P[h] := P_[h];
  end;
  
  setlength(ans, Q);
  for h := 0 to Q - 1 do begin
    ans[h] := N;
  end;
  setlength(queries, Q);
  setlength(uf, N);
  
  for phase := 0 to 1 do begin
    for i := 0 to C - 1 do begin
      if X[i] > Y[i] then begin
        swap(X[i], Y[i]);
      end;
    end;
    
    setlength(edges, C);
    edgesLen := 0;
    for i := 0 to C - 1 do begin
      if T[i] = 0 then begin
        edges[edgesLen] := i;
        edgesLen += 1;
      end;
    end;
    sort(edges, 0, edgesLen, @cmpEdge);
    setlength(link, C);
    for i := 0 to C - 1 do begin
      link[i] := C;
    end;
    for j := 0 to C - 1 do begin
      if T[j] = 1 then begin
        e := lowerBound(edges, 0, edgesLen, j, @cmpEdge) - 1;
        if e >= 0 then begin
          i := edges[e];
          if (X[i] = X[j]) and (Y[i] = Y[j]) then begin
            link[i] := j;
            link[j] := i;
          end;
        end;
      end;
    end;
    
    i := 0;
    while i < C do begin
      { solve for cable plans [i, j) }
      j := min(i + BLOCK_SIZE, C);
      queriesLen := 0;
      for h := 0 to Q - 1 do begin
        if (i <= W[h]) and (W[h] < j) then begin
          queries[queriesLen] := h;
          queriesLen += 1;
        end;
      end;
      sort(queries, 0, queriesLen, @cmpQuery);
      
      for u := 0 to N - 1 do begin
        uf[u] := -1;
      end;
      cntGlobal := 0;
      
      e := 0;
      for f := 0 to queriesLen - 1 do begin
        h := queries[f];
        { global edges }
        while (e < edgesLen) and (Y[edges[e]] <= P[h]) do begin
          if (edges[e] < i) and (j <= link[edges[e]]) then begin
            if connect(X[edges[e]], Y[edges[e]]) then begin
              cntGlobal += 1;
            end;
          end;
          e += 1;
        end;
        { path compression }
        for k := i to j - 1 do begin
          if ((T[k] = 0) and (k <= W[h]) and (W[h] < link[k])) or ((T[k] = 1) and (link[k] <= W[h]) and (W[h] < k)) then begin
            if Y[k] <= P[h] then begin
              root(X[k]);
              root(Y[k]);
            end;
          end;
        end;
        { backtracking }
        historyLen := 0;
        cntLocal := 0;
        for k := i to j - 1 do begin
          if ((T[k] = 0) and (k <= W[h]) and (W[h] < link[k])) or ((T[k] = 1) and (link[k] <= W[h]) and (W[h] < k)) then begin
            if Y[k] <= P[h] then begin
              if connectSaving(X[k], Y[k]) then begin
                cntLocal += 1;
              end;
            end;
          end;
        end;
        ans[h] -= cntGlobal;
        ans[h] -= cntLocal;
        for g := historyLen - 1 downto 0 do begin
          uf[historyKey[g]] := historyVal[g];
        end;
      end;
      
      i := j;
    end;
    
    for i := 0 to C - 1 do begin
      X[i] := N - 1 - X[i];
      Y[i] := N - 1 - Y[i];
    end;
    for h := 0 to Q - 1 do begin
      P[h] := N - 2 - P[h];
    end;
  end;
  
  exit(ans);
end;

end.

File: Olympiad/JOI/official/JOIOC/2018/code/2018-open-collapse.java
import static java.lang.Math.*;
import static java.util.Arrays.*;

public class collapse {
  final int BLOCK_SIZE = 333;
  final int MAX_HISTORY_LEN = 1000010;
  
  int[] simulateCollapse(int N, int[] T, int[] X, int[] Y, int[] W, int[] P) {
    this.N = N;
    this.C = T.length;
    this.Q = W.length;
    this.T = T;
    this.X = X;
    this.Y = Y;
    this.W = W;
    this.P = P;
    return run();
  }
  
  int N, C, Q;
  int[] T, X, Y, W, P;
  
  int[] run() {
    int[] ans = new int[Q];
    fill(ans, N);
    uf = new int[N];
    
    for (int phase = 0; phase < 2; ++phase) {
      for (int i = 0; i < C; ++i) {
        if (X[i] > Y[i]) {
          int tmp = X[i]; X[i] = Y[i]; Y[i] = tmp;
        }
      }
      
      int edgesLen = 0;
      for (int i = 0; i < C; ++i) if (T[i] == 0) {
        ++edgesLen;
      }
      Edge[] edges = new Edge[edgesLen];
      edgesLen = 0;
      for (int i = 0; i < C; ++i) if (T[i] == 0) {
        edges[edgesLen++] = new Edge(X[i], Y[i], i);
      }
      sort(edges);
      int[] link = new int[C];
      fill(link, C);
      for (int j = 0; j < C; ++j) if (T[j] == 1) {
        int e = ~binarySearch(edges, new Edge(X[j], Y[j], j)) - 1;
        if (e >= 0) {
          int i = edges[e].time;
          if (X[i] == X[j] && Y[i] == Y[j]) {
            link[i] = j;
            link[j] = i;
          }
        }
      }
      
      for (int i = 0; i < C; ) {
        // solve for cable plans [i, j)
        int j = min(i + BLOCK_SIZE, C);
        int queriesLen = 0;
        for (int h = 0; h < Q; ++h) {
          if (i <= W[h] && W[h] < j) {
            ++queriesLen;
          }
        }
        Query[] queries = new Query[queriesLen];
        queriesLen = 0;
        for (int h = 0; h < Q; ++h) {
          if (i <= W[h] && W[h] < j) {
            queries[queriesLen++] = new Query(W[h], P[h], h);
          }
        }
        sort(queries);
        
        fill(uf, -1);
        int cntGlobal = 0;
        
        int e = 0;
        for (Query q : queries) {
          // global edges
          for (; e < edgesLen && edges[e].y <= q.p; ++e) {
            if (edges[e].time < i && j <= link[edges[e].time]) {
              if (connect(edges[e].x, edges[e].y)) {
                ++cntGlobal;
              }
            }
          }
          // path compression
          for (int k = i; k < j; ++k) {
            if ((T[k] == 0 && k <= q.w && q.w < link[k]) || (T[k] == 1 && link[k] <= q.w && q.w < k)) {
              if (Y[k] <= q.p) {
                root(X[k]);
                root(Y[k]);
              }
            }
          }
          // backtracking
          historyLen = 0;
          int cntLocal = 0;
          for (int k = i; k < j; ++k) {
            if ((T[k] == 0 && k <= q.w && q.w < link[k]) || (T[k] == 1 && link[k] <= q.w && q.w < k)) {
              if (Y[k] <= q.p) {
                if (connectSaving(X[k], Y[k])) {
                  ++cntLocal;
                }
              }
            }
          }
          ans[q.id] -= cntGlobal;
          ans[q.id] -= cntLocal;
          for (int g = historyLen; g-- > 0; ) {
            uf[historyKey[g]] = historyVal[g];
          }
        }
        
        i = j;
      }
      
      for (int i = 0; i < C; ++i) {
        X[i] = N - 1 - X[i];
        Y[i] = N - 1 - Y[i];
      }
      for (int h = 0; h < Q; ++h) {
        P[h] = N - 2 - P[h];
      }
    }
    
    return ans;
  }
  
  int[] uf;
  int historyLen;
  int[] historyKey = new int[MAX_HISTORY_LEN];
  int[] historyVal = new int[MAX_HISTORY_LEN];
  int root(int u) {
    return (uf[u] < 0) ? u : (uf[u] = root(uf[u]));
  }
  boolean connect(int u, int v) {
    u = root(u);
    v = root(v);
    if (u == v) return false;
    if (uf[u] > uf[v]) { int tmp = u; u = v; v = tmp; }
    uf[u] += uf[v];
    uf[v] = u;
    return true;
  }
  void assignSaving(int key, int val) {
    if (uf[key] != val) {
      historyKey[historyLen] = key;
      historyVal[historyLen] = uf[key];
      ++historyLen;
      uf[key] = val;
    }
  }
  int rootSaving(int u) {
    if (uf[u] < 0) {
      return u;
    } else {
      assignSaving(u, rootSaving(uf[u]));
      return uf[u];
    }
  }
  boolean connectSaving(int u, int v) {
    u = rootSaving(u);
    v = rootSaving(v);
    if (u == v) return false;
    if (uf[u] > uf[v]) { int tmp = u; u = v; v = tmp; }
    assignSaving(u, uf[u] + uf[v]);
    assignSaving(v, u);
    return true;
  }
  
  class Edge implements Comparable<Edge> {
    int x, y, time;
    Edge(int x, int y, int time) {
      this.x = x;
      this.y = y;
      this.time = time;
    }
    public int compareTo(Edge o) {
      return (y != o.y) ? (y - o.y) : (x != o.x) ? (x - o.x) : (time - o.time);
    }
  }
  
  class Query implements Comparable<Query> {
    int w, p, id;
    Query(int w, int p, int id) {
      this.w = w;
      this.p = p;
      this.id = id;
    }
    public int compareTo(Query o) {
      return (p != o.p) ? (p - o.p) : (id - o.id);
    }
  }
}
